File: javascript/nodejs/queue_processing/cache-manager.js
/**
 * Cache Manager with LRU Eviction Policy
 * 
 * Provides efficient in-memory caching with automatic eviction based on:
 * - Least Recently Used (LRU) policy
 * - Maximum size limits
 */

const { LRUCache } = require('lru-cache');

class CacheManager {
  /**
   * Initialize cache manager
   * @param {number} maxSize - Maximum number of items in cache
   * @param {number} ttl - Time-to-live in milliseconds (optional)
   */
  constructor(maxSize = 10000, ttl = null) {
    this.maxSize = maxSize;
    this.ttl = ttl;
    
    const options = {
      max: maxSize
    };
    
    if (ttl) {
      options.ttl = ttl;
    }
    
    this.cache = new LRUCache(options);
    this.hits = 0;
    this.misses = 0;
  }

  /**
   * Retrieve item from cache
   * @param {string} key - Cache key
   * @returns {*} Cached value or undefined if not found
   */
  get(key) {
    const value = this.cache.get(key);
    
    if (value !== undefined) {
      this.hits++;
      return value;
    }
    
    this.misses++;
    return undefined;
  }

  /**
   * Add or update item in cache
   * @param {string} key - Cache key
   * @param {*} value - Value to cache
   */
  set(key, value) {
    this.cache.set(key, value);
  }

  /**
   * Remove item from cache
   * @param {string} key - Cache key
   * @returns {boolean} True if item was removed
   */
  delete(key) {
    return this.cache.delete(key);
  }

  /**
   * Clear all items from cache
   */
  clear() {
    this.cache.clear();
    this.hits = 0;
    this.misses = 0;
  }

  /**
   * Get current cache size
   * @returns {number} Current size
   */
  size() {
    return this.cache.size;
  }

  /**
   * Get cache statistics
   * @returns {Object} Cache statistics
   */
  getStats() {
    const totalRequests = this.hits + this.misses;
    const hitRate = totalRequests > 0 ? (this.hits / totalRequests * 100) : 0;

    return {
      size: this.cache.size,
      maxSize: this.maxSize,
      hits: this.hits,
      misses: this.misses,
      hitRate: `${hitRate.toFixed(2)}%`,
      totalRequests
    };
  }
}

class MarketDataCache extends CacheManager {
  /**
   * Specialized cache for market data with symbol-based keys
   */
  
  /**
   * Get order by ID
   * @param {string} orderId - Order ID
   * @returns {Object|undefined} Order data
   */
  getOrder(orderId) {
    return this.get(`order:${orderId}`);
  }

  /**
   * Cache order data
   * @param {string} orderId - Order ID
   * @param {Object} orderData - Order data
   */
  setOrder(orderId, orderData) {
    this.set(`order:${orderId}`, orderData);
  }

  /**
   * Get cached price for symbol
   * @param {string} symbol - Stock symbol
   * @returns {number|undefined} Price
   */
  getSymbolPrice(symbol) {
    return this.get(`price:${symbol}`);
  }

  /**
   * Cache symbol price
   * @param {string} symbol - Stock symbol
   * @param {number} price - Price
   */
  setSymbolPrice(symbol, price) {
    this.set(`price:${symbol}`, price);
  }
}

// Example usage
if (require.main === module) {
  console.log('=== Basic Cache Operations ===');
  const cache = new CacheManager(5);

  cache.set('key1', 'value1');
  cache.set('key2', 'value2');
  cache.set('key3', 'value3');

  console.log('Get key1:', cache.get('key1'));
  console.log('Get key2:', cache.get('key2'));
  console.log('Get missing:', cache.get('missing'));

  console.log('\nCache stats:', cache.getStats());

  // LRU eviction demo
  console.log('\n=== LRU Eviction Demo ===');
  for (let i = 0; i < 10; i++) {
    cache.set(`key${i}`, `value${i}`);
  }

  console.log(`Cache size: ${cache.size()} (max: ${cache.maxSize})`);
  console.log('Oldest keys evicted, newest keys remain');

  // Market data cache
  console.log('\n=== Market Data Cache ===');
  const marketCache = new MarketDataCache(1000);

  marketCache.setOrder('ORD001', {
    symbol: 'AAPL',
    quantity: 100,
    price: 150.50,
    side: 'BUY'
  });

  marketCache.setSymbolPrice('AAPL', 150.75);

  console.log('Order:', marketCache.getOrder('ORD001'));
  console.log('Price:', marketCache.getSymbolPrice('AAPL'));
  console.log('\nStats:', marketCache.getStats());
}

module.exports = { CacheManager, MarketDataCache };

File: python/samples/queue_processing/cache_manager.py
"""
Cache Manager with LRU Eviction Policy

Provides efficient in-memory caching with automatic eviction based on:
- Least Recently Used (LRU) policy
- Maximum size limits
- Memory usage thresholds
"""

from collections import OrderedDict
from typing import Any, Optional
import time


class CacheManager:
    """
    Thread-safe LRU cache implementation with memory management
    
    Features:
    - LRU eviction policy
    - Configurable max size
    - Access time tracking
    - Hit/miss statistics
    """
    
    def __init__(self, max_size: int = 10000, ttl_seconds: Optional[int] = None):
        """
        Initialize cache manager
        
        Args:
            max_size: Maximum number of items in cache
            ttl_seconds: Time-to-live for cache entries (None = no expiration)
        """
        self.max_size = max_size
        self.ttl_seconds = ttl_seconds
        self.cache = OrderedDict()
        self.timestamps = {}
        self.hits = 0
        self.misses = 0
        
    def get(self, key: str) -> Optional[Any]:
        """
        Retrieve item from cache
        
        Args:
            key: Cache key
            
        Returns:
            Cached value or None if not found/expired
        """
        # Check if key exists
        if key not in self.cache:
            self.misses += 1
            return None
            
        # Check if expired
        if self.ttl_seconds and self._is_expired(key):
            self._evict(key)
            self.misses += 1
            return None
            
        # Move to end (mark as recently used)
        self.cache.move_to_end(key)
        self.hits += 1
        return self.cache[key]
        
    def set(self, key: str, value: Any) -> None:
        """
        Add or update item in cache
        
        Args:
            key: Cache key
            value: Value to cache
        """
        # If key exists, update and move to end
        if key in self.cache:
            self.cache.move_to_end(key)
        else:
            # Check if we need to evict
            if len(self.cache) >= self.max_size:
                self._evict_lru()
                
        self.cache[key] = value
        self.timestamps[key] = time.time()
        
    def delete(self, key: str) -> bool:
        """
        Remove item from cache
        
        Args:
            key: Cache key
            
        Returns:
            True if item was removed, False if not found
        """
        if key in self.cache:
            del self.cache[key]
            del self.timestamps[key]
            return True
        return False
        
    def clear(self) -> None:
        """Clear all items from cache"""
        self.cache.clear()
        self.timestamps.clear()
        self.hits = 0
        self.misses = 0
        
    def size(self) -> int:
        """Get current cache size"""
        return len(self.cache)
        
    def get_stats(self) -> dict:
        """
        Get cache statistics
        
        Returns:
            Dictionary with cache statistics
        """
        total_requests = self.hits + self.misses
        hit_rate = (self.hits / total_requests * 100) if total_requests > 0 else 0
        
        return {
            "size": len(self.cache),
            "max_size": self.max_size,
            "hits": self.hits,
            "misses": self.misses,
            "hit_rate": f"{hit_rate:.2f}%",
            "total_requests": total_requests
        }
        
    def _is_expired(self, key: str) -> bool:
        """Check if cache entry has expired"""
        if not self.ttl_seconds:
            return False
        timestamp = self.timestamps.get(key)
        if not timestamp:
            return True
        return (time.time() - timestamp) > self.ttl_seconds
        
    def _evict_lru(self) -> None:
        """Evict least recently used item"""
        if self.cache:
            # Remove first item (least recently used)
            key = next(iter(self.cache))
            self._evict(key)
            
    def _evict(self, key: str) -> None:
        """Evict specific key"""
        if key in self.cache:
            del self.cache[key]
        if key in self.timestamps:
            del self.timestamps[key]


class MarketDataCache(CacheManager):
    """
    Specialized cache for market data with symbol-based keys
    """
    
    def get_order(self, order_id: str) -> Optional[dict]:
        """Get order by ID"""
        return self.get(f"order:{order_id}")
        
    def set_order(self, order_id: str, order_data: dict) -> None:
        """Cache order data"""
        self.set(f"order:{order_id}", order_data)
        
    def get_symbol_price(self, symbol: str) -> Optional[float]:
        """Get cached price for symbol"""
        return self.get(f"price:{symbol}")
        
    def set_symbol_price(self, symbol: str, price: float) -> None:
        """Cache symbol price"""
        self.set(f"price:{symbol}", price)


# Example usage
if __name__ == "__main__":
    # Basic cache usage
    cache = CacheManager(max_size=5)
    
    print("=== Basic Cache Operations ===")
    cache.set("key1", "value1")
    cache.set("key2", "value2")
    cache.set("key3", "value3")
    
    print(f"Get key1: {cache.get('key1')}")
    print(f"Get key2: {cache.get('key2')}")
    print(f"Get missing: {cache.get('missing')}")
    
    print(f"\nCache stats: {cache.get_stats()}")
    
    # LRU eviction demo
    print("\n=== LRU Eviction Demo ===")
    for i in range(10):
        cache.set(f"key{i}", f"value{i}")
        
    print(f"Cache size: {cache.size()} (max: {cache.max_size})")
    print(f"Oldest keys evicted, newest keys remain")
    
    # Market data cache
    print("\n=== Market Data Cache ===")
    market_cache = MarketDataCache(max_size=1000)
    
    market_cache.set_order("ORD001", {
        "symbol": "AAPL",
        "quantity": 100,
        "price": 150.50,
        "side": "BUY"
    })
    
    market_cache.set_symbol_price("AAPL", 150.75)
    
    print(f"Order: {market_cache.get_order('ORD001')}")
    print(f"Price: {market_cache.get_symbol_price('AAPL')}")
    print(f"\nStats: {market_cache.get_stats()}")
